File: src/server/repository.ts
import { sql } from "./db";
import type { BoundsInput, CameraInput, CameraScenario, Region } from "../shared/types";

type DbRegion = Region;
type DbScenario = CameraScenario;

export async function listRegions(): Promise<Region[]> {
  return sql<DbRegion[]>`select id::int, name, north, south, east, west, created_at::text from regions order by created_at desc`;
}

export async function getRegion(id: number): Promise<Region> {
  const rows = await sql<DbRegion[]>`
    select id::int, name, north, south, east, west, created_at::text
    from regions
    where id = ${id}
  `;
  const region = rows[0];
  if (!region) throw new HttpError(404, "Region not found.");
  return region;
}

export async function createRegion(name: string, bounds: BoundsInput): Promise<Region> {
  const rows = await sql<DbRegion[]>`
    insert into regions (name, north, south, east, west, bbox)
    values (
      ${name},
      ${bounds.north},
      ${bounds.south},
      ${bounds.east},
      ${bounds.west},
      ST_MakeEnvelope(${bounds.west}, ${bounds.south}, ${bounds.east}, ${bounds.north}, 4326)
    )
    returning id::int, name, north, south, east, west, created_at::text
  `;
  return rows[0]!;
}

export async function listScenarios(regionId: number): Promise<CameraScenario[]> {
  return sql<DbScenario[]>`
    select cs.id::int, cs.region_id::int, cs.name, cs.source, cs.created_at::text, count(c.id)::int as camera_count
    from camera_sets cs
    left join cameras c on c.camera_set_id = cs.id
    where cs.region_id = ${regionId}
    group by cs.id
    order by cs.created_at desc
  `;
}

export async function getScenario(id: number): Promise<CameraScenario> {
  const rows = await sql<DbScenario[]>`
    select id::int, region_id::int, name, source, created_at::text
    from camera_sets
    where id = ${id}
  `;
  const scenario = rows[0];
  if (!scenario) throw new HttpError(404, "Scenario not found.");
  return scenario;
}

export async function createScenario(
  regionId: number,
  name: string,
  source: CameraScenario["source"],
  cameras: CameraInput[],
): Promise<CameraScenario> {
  return sql.begin(async (tx) => {
    const scenarios = await tx<DbScenario[]>`
      insert into camera_sets (region_id, name, source)
      values (${regionId}, ${name}, ${source})
      returning id::int, region_id::int, name, source, created_at::text
    `;
    const scenario = scenarios[0]!;

    for (const camera of cameras) {
      await tx`
        insert into cameras (camera_set_id, camera, lat, long, orientation_deg, fov_deg, range_m, geom)
        values (
          ${scenario.id},
          ${camera.camera},
          ${camera.lat},
          ${camera.long},
          ${camera.orientation_deg},
          ${camera.fov_deg},
          ${camera.range_m},
          ST_SetSRID(ST_MakePoint(${camera.long}, ${camera.lat}), 4326)
        )
      `;
    }

    return { ...scenario, camera_count: cameras.length };
  });
}

export async function getCameras(scenarioId: number): Promise<Array<CameraInput & { id: number }>> {
  return sql<Array<CameraInput & { id: number }>>`
    select id::int, camera, lat, long, orientation_deg, fov_deg, range_m
    from cameras
    where camera_set_id = ${scenarioId}
    order by id
  `;
}

export async function getCameraFeatureCollection(scenarioId: number): Promise<GeoJSON.FeatureCollection> {
  const rows = await sql<Array<{ feature: GeoJSON.Feature }>>`
    select jsonb_build_object(
      'type', 'Feature',
      'properties', jsonb_build_object(
        'id', id::int,
        'camera', camera,
        'orientation_deg', orientation_deg,
        'fov_deg', fov_deg,
        'range_m', range_m
      ),
      'geometry', ST_AsGeoJSON(geom)::jsonb
    ) as feature
    from cameras
    where camera_set_id = ${scenarioId}
    order by id
  `;
  return featureCollection(rows.map((row) => row.feature));
}

export async function getBuildingFeatureRows(regionId: number): Promise<Array<{ id: number; feature: GeoJSON.Feature }>> {
  return sql<Array<{ id: number; feature: GeoJSON.Feature }>>`
    select id::int, jsonb_build_object(
      'type', 'Feature',
      'properties', jsonb_build_object('id', id::int, 'tags', tags, 'source_osm_type', source_osm_type, 'source_osm_id', source_osm_id),
      'geometry', ST_AsGeoJSON(geom)::jsonb
    ) as feature
    from buildings
    where region_id = ${regionId}
    order by id
  `;
}

export async function getBuildings(regionId: number): Promise<GeoJSON.FeatureCollection> {
  const rows = await getBuildingFeatureRows(regionId);
  return featureCollection(rows.map((row) => row.feature));
}

export async function upsertBuildings(
  regionId: number,
  buildings: Array<{ sourceType: string; sourceId: string; tags: Record<string, unknown>; geometry: GeoJSON.MultiPolygon }>,
) {
  for (const building of buildings) {
    await sql`
      insert into buildings (region_id, source_osm_type, source_osm_id, tags, geom)
      values (
        ${regionId},
        ${building.sourceType},
        ${building.sourceId},
        ${JSON.stringify(building.tags)}::jsonb,
        ST_SetSRID(ST_GeomFromGeoJSON(${JSON.stringify(building.geometry)}), 4326)
      )
      on conflict (region_id, source_osm_type, source_osm_id)
      do update set tags = excluded.tags, geom = excluded.geom, imported_at = now()
    `;
  }
}

export async function createAnalysisRun(input: {
  regionId: number;
  scenarioId: number;
  groundCellSizeM: number;
  coverage: GeoJSON.FeatureCollection;
  wallNormals: GeoJSON.FeatureCollection;
  wallBlindspots: GeoJSON.FeatureCollection;
  groundBlindspots: GeoJSON.FeatureCollection;
}) {
  return sql.begin(async (tx) => {
    const runs = await tx<Array<{ id: number; scenario_id: number; created_at: string; ground_cell_size_m: number }>>`
      insert into analysis_runs (camera_set_id, region_id, ground_cell_size_m, coverage_geojson, wall_normals_geojson)
      values (
        ${input.scenarioId},
        ${input.regionId},
        ${input.groundCellSizeM},
        ${JSON.stringify(input.coverage)}::jsonb,
        ${JSON.stringify(input.wallNormals)}::jsonb
      )
      returning id::int, camera_set_id::int as scenario_id, created_at::text, ground_cell_size_m
    `;
    const run = runs[0]!;

    for (const feature of input.wallBlindspots.features) {
      await tx`
        insert into wall_blindspots (analysis_run_id, building_id, length_m, geom)
        values (
          ${run.id},
          ${(feature.properties?.building_id as number | undefined) ?? null},
          ${Number(feature.properties?.length_m ?? 0)},
          ST_SetSRID(ST_GeomFromGeoJSON(${JSON.stringify(feature.geometry)}), 4326)
        )
      `;
    }

    for (const feature of input.groundBlindspots.features) {
      await tx`
        insert into ground_blindspots (analysis_run_id, geom)
        values (${run.id}, ST_SetSRID(ST_GeomFromGeoJSON(${JSON.stringify(feature.geometry)}), 4326))
      `;
    }

    return run;
  });
}

export function featureCollection(features: GeoJSON.Feature[]): GeoJSON.FeatureCollection {
  return { type: "FeatureCollection", features };
}

export class HttpError extends Error {
  constructor(
    readonly status: number,
    message: string,
  ) {
    super(message);
  }
}

File: blindspot/lib/repository.py
import json
from contextlib import contextmanager
from typing import Any

from django.db import connection, transaction

from .geo import feature_collection
from .errors import HttpError


@contextmanager
def cursor():
    with connection.cursor() as cur:
        yield cur


def rows_to_dicts(cur) -> list[dict[str, Any]]:
    columns = [column[0] for column in cur.description]
    return [dict(zip(columns, row, strict=False)) for row in cur.fetchall()]


def one_or_404(rows: list[dict], message: str) -> dict:
    if not rows:
        raise HttpError(404, message)
    return rows[0]


def list_regions() -> list[dict]:
    with cursor() as cur:
        cur.execute(
            """
            select id::int, name, north, south, east, west, created_at::text
            from regions
            order by case when name like 'Seeded %%' then 0 else 1 end, id asc
            """
        )
        return rows_to_dicts(cur)


def get_region(region_id: int) -> dict:
    with cursor() as cur:
        cur.execute(
            "select id::int, name, north, south, east, west, created_at::text from regions where id = %s",
            [region_id],
        )
        return one_or_404(rows_to_dicts(cur), "Region not found.")


def create_region(name: str, bounds: dict) -> dict:
    with cursor() as cur:
        cur.execute(
            """
            insert into regions (name, north, south, east, west, bbox)
            values (%s, %s, %s, %s, %s, ST_MakeEnvelope(%s, %s, %s, %s, 4326))
            returning id::int, name, north, south, east, west, created_at::text
            """,
            [
                name,
                bounds["north"],
                bounds["south"],
                bounds["east"],
                bounds["west"],
                bounds["west"],
                bounds["south"],
                bounds["east"],
                bounds["north"],
            ],
        )
        return rows_to_dicts(cur)[0]


def list_scenarios(region_id: int) -> list[dict]:
    with cursor() as cur:
        cur.execute(
            """
            select cs.id::int, cs.region_id::int, cs.name, cs.source, cs.created_at::text, count(c.id)::int as camera_count
            from camera_sets cs
            left join cameras c on c.camera_set_id = cs.id
            where cs.region_id = %s
            group by cs.id
            order by cs.created_at desc
            """,
            [region_id],
        )
        return rows_to_dicts(cur)


def get_scenario(scenario_id: int) -> dict:
    with cursor() as cur:
        cur.execute(
            "select id::int, region_id::int, name, source, created_at::text from camera_sets where id = %s",
            [scenario_id],
        )
        return one_or_404(rows_to_dicts(cur), "Scenario not found.")


def create_scenario(region_id: int, name: str, source: str, cameras: list[dict]) -> dict:
    with transaction.atomic():
        with cursor() as cur:
            cur.execute(
                """
                insert into camera_sets (region_id, name, source)
                values (%s, %s, %s)
                returning id::int, region_id::int, name, source, created_at::text
                """,
                [region_id, name, source],
            )
            scenario = rows_to_dicts(cur)[0]
            for camera in cameras:
                cur.execute(
                    """
                    insert into cameras (camera_set_id, camera, lat, long, orientation_deg, fov_deg, range_m, geom)
                    values (%s, %s, %s, %s, %s, %s, %s, ST_SetSRID(ST_MakePoint(%s, %s), 4326))
                    """,
                    [
                        scenario["id"],
                        camera["camera"],
                        camera["lat"],
                        camera["long"],
                        camera["orientation_deg"],
                        camera["fov_deg"],
                        camera["range_m"],
                        camera["long"],
                        camera["lat"],
                    ],
                )
            scenario["camera_count"] = len(cameras)
            return scenario


def get_cameras(scenario_id: int) -> list[dict]:
    with cursor() as cur:
        cur.execute(
            """
            select id::int, camera, lat, long, orientation_deg, fov_deg, range_m
            from cameras
            where camera_set_id = %s
            order by id
            """,
            [scenario_id],
        )
        return rows_to_dicts(cur)


def get_camera_feature_collection(scenario_id: int) -> dict:
    with cursor() as cur:
        cur.execute(
            """
            select jsonb_build_object(
              'type', 'Feature',
              'properties', jsonb_build_object(
                'id', id::int,
                'camera', camera,
                'orientation_deg', orientation_deg,
                'fov_deg', fov_deg,
                'range_m', range_m
              ),
              'geometry', ST_AsGeoJSON(geom)::jsonb
            ) as feature
            from cameras
            where camera_set_id = %s
            order by id
            """,
            [scenario_id],
        )
        return feature_collection([row["feature"] for row in rows_to_dicts(cur)])


def get_building_feature_rows(region_id: int) -> list[dict]:
    with cursor() as cur:
        cur.execute(
            """
            select id::int, jsonb_build_object(
              'type', 'Feature',
              'properties', jsonb_build_object('id', id::int, 'tags', tags, 'source_osm_type', source_osm_type, 'source_osm_id', source_osm_id),
              'geometry', ST_AsGeoJSON(geom)::jsonb
            ) as feature
            from buildings
            where region_id = %s
            order by id
            """,
            [region_id],
        )
        return rows_to_dicts(cur)


def get_buildings(region_id: int) -> dict:
    return feature_collection([row["feature"] for row in get_building_feature_rows(region_id)])


def upsert_buildings(region_id: int, buildings: list[dict]) -> None:
    with cursor() as cur:
        for building in buildings:
            cur.execute(
                """
                insert into buildings (region_id, source_osm_type, source_osm_id, tags, geom)
                values (%s, %s, %s, %s::jsonb, ST_SetSRID(ST_GeomFromGeoJSON(%s), 4326))
                on conflict (region_id, source_osm_type, source_osm_id)
                do update set tags = excluded.tags, geom = excluded.geom, imported_at = now()
                """,
                [
                    region_id,
                    building["sourceType"],
                    building["sourceId"],
                    json.dumps(building["tags"]),
                    json.dumps(building["geometry"]),
                ],
            )


def create_analysis_run(data: dict) -> dict:
    with transaction.atomic():
        with cursor() as cur:
            cur.execute(
                """
                insert into analysis_runs (camera_set_id, region_id, ground_cell_size_m, coverage_geojson, wall_normals_geojson)
                values (%s, %s, %s, %s::jsonb, %s::jsonb)
                returning id::int, camera_set_id::int as scenario_id, created_at::text, ground_cell_size_m
                """,
                [
                    data["scenarioId"],
                    data["regionId"],
                    data["groundCellSizeM"],
                    json.dumps(data["coverage"]),
                    json.dumps(data["wallNormals"]),
                ],
            )
            run = rows_to_dicts(cur)[0]
            for feature in data["wallBlindspots"]["features"]:
                cur.execute(
                    """
                    insert into wall_blindspots (analysis_run_id, building_id, length_m, geom)
                    values (%s, %s, %s, ST_SetSRID(ST_GeomFromGeoJSON(%s), 4326))
                    """,
                    [
                        run["id"],
                        feature.get("properties", {}).get("building_id"),
                        float(feature.get("properties", {}).get("length_m", 0)),
                        json.dumps(feature["geometry"]),
                    ],
                )
            for feature in data["groundBlindspots"]["features"]:
                cur.execute(
                    """
                    insert into ground_blindspots (analysis_run_id, geom)
                    values (%s, ST_SetSRID(ST_GeomFromGeoJSON(%s), 4326))
                    """,
                    [run["id"], json.dumps(feature["geometry"])],
                )
            return run
